nova_vm/src/ecmascript/abstract_operations/operations_on_objects.rs

```diff
@@ -414,6 +415,58 @@ pub(crate) fn length_of_array_like(agent: &mut Agent, obj: Object) -> JsResult<i
     to_length(agent, property)
 }
 
+/// ### [7.3.19 CreateListFromArrayLike ( obj [ , elementTypes ] )](https://tc39.es/ecma262/#sec-createlistfromarraylike)
+///
+/// The abstract operation CreateListFromArrayLike takes argument obj (an ECMAScript language value)
+/// and optional argument elementTypes (a List of names of ECMAScript Language Types) and returns
+/// either a normal completion containing a List of ECMAScript language values or a throw
+/// completion. It is used to create a List value whose elements are provided by the indexed
+/// properties of obj. elementTypes contains the names of ECMAScript Language Types that are allowed
+/// for element values of the List that is created.
+///
+/// NOTE: This implementation doesn't yet support `elementTypes`.
+pub(crate) fn create_list_from_array_like(agent: &mut Agent, obj: Value) -> JsResult<Vec<Value>> {
+    // TODO: Are there any cases where we might want to return a `Cow<[Value]>` (borrowing the
```

Comment:
Do you think I should keep the TODO?

Reply:
I can think of two places where it would be conditionally useful:
1. **On the condition** that we add variable length argument support to function call signatures (behaviours), ie. we know exactly up to how many arguments the function will ever access, then we could always use a CoW value for function calls; the only copy would happen from the slice into the proper passing registers.
2. **On the condition** that we make sure to never drop heap vectors while JS may be running, and we track literal objects/arrays that are never assigned, eg. the array in `apply(null, [arg0])`, then we could use the backing store slice directly here.

The first one is fully possible for and kind of in the plans for builtin functions. It's also possible for ECMAScript functions but would require careful analysis of the AST/bytecode and would be impure (you could see that we're doing this by using a breakpoint in a function that you call with more arguments than it would every normally expect, and then accessing `arguments`).

The second one is... in some ways easier: It's fairly easy to detect these sorts of object and array literals. But then we'd also need to guarantee the "no heap vector drops while code is running" in the general case, and that becomes a bit annoying. Although it's not too bad either. The biggest issue would probably be the "is this object/array still unassigned?"

But that does give me a potentially interesting idea... What if we have extra `ObjectLiteral` and `ArrayLiteral` Value variants that would only appear in these very specific cases, and would be exactly equivalent to Objects and Arrays, but if we eg. see a call like `[a, b].includes(candidate)` then the `Array.prototype.includes` call could check that it's got an ArrayLiteral as `this` and could delete it immediately at the end of the call... Well, maybe interesting. Maybe not. Not for today, anyhow.

Reply:
So that is to say: I don't have a strong opinion either way. I'd _maybe_ not leave it, or I'd maybe leave it but mentioning more in depth the places where this would be useful and what the conditions / caveats around it are.

Reply:
I think I'll remove the TODO, but also open an issue about that.